internal/controller/volumereplicationgroup_controller.go

```diff
@@ -800,41 +795,267 @@ func (v *VRGInstance) separatePVCsUsingVRGStatus(pvcList *corev1.PersistentVolum
-		}
+		if len(peerClasses) != 0 {
+			if _, ok := storageClass.Labels[StorageIDLabel]; !ok {
+				return fmt.Errorf("storageID label not found in storageClass for sync PVC %s", pvc.Name)
+			}
 
-		replicationClassMatchFound := false
+			peerClass := v.findPeerClassMatchingSC(peerClasses, *scName)
+			if peerClass == nil {
+				return fmt.Errorf("peerClass matching storageClass %s not found for sync PVC", *scName)
+			}
 
-		for _, replicationClass := range v.replClassList.Items {
-			if storageClass.Provisioner == replicationClass.Spec.Provisioner {
-				v.volRepPVCs = append(v.volRepPVCs, *pvc)
-				replicationClassMatchFound = true
+			if len(peerClass.StorageID) != 1 {
+				return fmt.Errorf("no StorageID found in sync PeerClass")
+			}
 
-				break
+			if storageClass.Labels[StorageIDLabel] != peerClass.StorageID[0] {
+				return fmt.Errorf("storageId %s not present in peerClass", storageClass.Labels[StorageIDLabel])
 			}
 		}
+	}
 
-		if !replicationClassMatchFound {
-			v.volSyncPVCs = append(v.volSyncPVCs, *pvc)
+	v.volRepPVCs = make([]corev1.PersistentVolumeClaim, len(pvcList.Items))
+	total := copy(v.volRepPVCs, pvcList.Items)
+	v.log.Info("Found PersistentVolumeClaims", "count", total)
+
+	return nil
+}
+
+//nolint:gocognit,cyclop
+func (v *VRGInstance) separatePVCsUsingStorageClassProvisioner(pvcList *corev1.PersistentVolumeClaimList) error {
+	peerClasses := v.instance.Spec.Async.PeerClasses
+
+	for idx := range pvcList.Items {
+		pvc := &pvcList.Items[idx]
+		scName := pvc.Spec.StorageClassName
+
+		storageClass, err := v.validateAndGetStorageClass(scName, pvc)
+		if err != nil {
+			return err
+		}
+
+		v.log.Info(fmt.Sprintf("storageClass %s validated", storageClass.Name))
+
+		switch len(v.instance.Spec.Async.PeerClasses) {
+		case 0:
+			v.log.Info("validate using only sc provisioner")
+
+			replicationClassMatchFound := false
+
+			for _, replicationClass := range v.replClassList.Items {
+				if storageClass.Provisioner == replicationClass.Spec.Provisioner {
+					v.volRepPVCs = append(v.volRepPVCs, *pvc)
+					replicationClassMatchFound = true
+
+					break
+				}
+			}
+
+			if !replicationClassMatchFound {
+				v.volSyncPVCs = append(v.volSyncPVCs, *pvc)
+			}
+		default:
+			v.log.Info("filtering using peerClasses")
+
+			replicationClass, err := v.findReplicationClassUsingPeerClasses(peerClasses, storageClass, pvc)
+			if err != nil {
+				return err
+			}
+
+			if replicationClass == nil && !v.instance.Spec.VolSync.Disabled {
+				// ignoring volsnapClass till we store it in the future
+				if _, err := v.findVolSnapClass(storageClass, pvc); err != nil {
+					return err
+				}
+			}
 		}
 	}
 
 	v.log.Info(fmt.Sprintf("Found %d PVCs targeted for VolRep and %d targeted for VolSync",
 		len(v.volRepPVCs), len(v.volSyncPVCs)))
 
+	if !(len(pvcList.Items) == (len(v.volRepPVCs) + len(v.volSyncPVCs))) {
+		return fmt.Errorf("no PVCs are procted")
```

Comment:
Today, we prioritized protecting PVCs with VolRep first and defaulted to VolSync if no VRC is available. This change, however, is making it to either protect all PVCs or none. I am not sure if that's correct... I am trying to think if there is a case where it is still ok to start protecting n-1 PVCs until all are protected...